File: AddressCardOC/AddressCard/AddressBook.m
//
//  AddressBook.m
//  AddressCard
//
//  Created by LeonTse on 15/6/5.
//  Copyright (c) 2015年 LeonTse. All rights reserved.
//

#import "AddressBook.h"

@implementation AddressBook
@synthesize bookName, book;

//
-(instancetype) initWithName:(NSString *)name
{
    self = [super init];
    if (self) {
        bookName = [NSString stringWithString:name];
        book = [NSMutableArray array];
    }
    
    return self;
}
-(instancetype) init
{
    return [self initWithName:@"NoName"];
}

-(void) addCard:(AddressCard *)theCard
{
    [book addObject: theCard];
}
-(int) entries
{
    return (int)[book count];
}

-(void) list
{
    NSLog(@"============ Contents of: %@ =========", bookName);
    for (AddressCard *theCard in book) {
        NSLog(@"%-20s      %-32s", [theCard.name UTF8String], [theCard.email UTF8String]);
    }
    NSLog(@"=============================================");
}

-(AddressCard *) lookup:(NSString *)theName
{
//    NSRange subRange;
//    NSIndexSet *index;
//    for (AddressCard *add in book)
//    {
//        subRange = [[add name] rangeOfString: theName];
//        if (subRange.location != NSNotFound)
//        {
//            [index ]
//        }
//    }
//    
//    
//    if (result != NSNotFound) {
//        return book[result];
//    }
//    else
//        return nil;
    for (AddressCard *nextCard in book)
        if ( [nextCard.name caseInsensitiveCompare:theName] ==NSOrderedSame)
            return nextCard;
    return nil;
}

-(NSArray *) lookupAll: (NSString *) theName
{
    NSIndexSet *indexSet = [book indexesOfObjectsPassingTest:
                          ^(id obj, NSUInteger idx, BOOL *stop)
    {
        if ([[obj name] caseInsensitiveCompare: theName] == NSOrderedSame)
            return YES; // found a match, keep going
        else
            return NO; // keep looking
    } ];
    // Return the result
    //return result;
    if (indexSet != nil) {
        NSArray *result = [book objectsAtIndexes:indexSet];
        return result;
    }
    else
        return nil;
    
}

-(void) removeCard:(AddressCard *)theCard
{
    [book removeObjectIdenticalTo:theCard];
}

-(void) sort
{
    [book sortUsingComparator:^(id obj1, id obj2){
        return [[obj1 name] compare:[obj2 name]];
    }];
    //   [book sortUsingSelector:@selector(compareNames:)];
}
@end

File: AddressCardSwift/AddressCard/AddressBook.swift
//
//  AddressBook.swift
//  AddressCard
//
//  Created by LeonTse on 2017/4/1.
//  Copyright © 2017年 LeonTse. All rights reserved.
//

//
//  AddressBook.m
//  AddressCard
//
//  Created by LeonTse on 15/6/5.
//  Copyright (c) 2015年 LeonTse. All rights reserved.
//

import Foundation

class AddressBook : NSObject
{
    var bookName:String!
    var book:NSMutableArray
    
    init(name:String)
    {
        bookName = name
        book = NSMutableArray.init()
    }
    
    override convenience init()
    {
        self.init(name:"NoName")
    }
    @objc(addCard:)
    public func addCard(theCard:AddressCard)
    {
        book.insert(theCard, at: book.count)
    }
    
    func entries()->Int
    {
        return book.count
    }
    
    func list()
    {
        print("============ Contents of: \(bookName) =========")
        for theCard in book
        {
            print(String(format:"%-20s",(theCard as! AddressCard).name)+"      "+String(format:"%-32s",(theCard as! AddressCard).email))
        }
        print("=============================================")
    }
    
    func lookup(theName :String) -> AddressCard?
    {
        for nextCard  in book
        {
            if (nextCard as! AddressCard).name.caseInsensitiveCompare(theName) == ComparisonResult.orderedSame
            {
                return nextCard as? AddressCard
            }
        }
        return nil
    }
    @objc(lookupAll:)
    public func lookupAll(theName: String) -> Array<Any>?
    {
        var indexSet = book.indexesOfObjects(options: NSEnumerationOptions.concurrent, passingTest: { (obj:Any, idx:Int, stop:UnsafeMutablePointer<ObjCBool>) -> Bool in
            
                return (obj as! AddressCard).name.caseInsensitiveCompare(theName) == ComparisonResult.orderedSame
            
        })
        if !indexSet.isEmpty
        {
            var result = book.objects(at: indexSet)
            return result
        }
        else
        {
            return nil
        }
    }
    
    func removeCard(theCard:AddressCard)
    {
        book.removeObject(identicalTo:theCard)
    }
    
    func sort()
    {
        book.sort(options: NSSortOptions.concurrent) { (obj1:Any, obj2: Any) -> ComparisonResult in
            
                return (obj1 as! AddressCard).name.compare((obj2 as! AddressCard).name)
            
            
        }
    }

}
